src/clj/rems/api/catalogue_items.clj

```diff
@@ -102,18 +102,20 @@
         (ok it)
         (not-found-json-response)))
 
-    (POST "/create" []
+    (POST "/create" request
       :summary "Create a new catalogue item"
       :roles +admin-write-roles+
       :body [command CreateCatalogueItemCommand]
       :return CreateCatalogueItemResponse
+      (extended-logging request command)
```

Comment:
Grand! I'm happy that my prediction that, a more experienced Clojure developer could find a neater solution, was fulfilled!

Side note: I assumed that `:params` related to url query params, and glossed over it. Maybe it would improve readability to avoid a name with an overlap with a closely related domain, maybe `:vars`?

Reply:
Unfortunately, I do not get any content with a more general strategy

```
2023-09-20 13:05:00,669 [jetty-183 rqc:1 owner oQdVKBYp] INFO  rems.middleware - req > :put /api/forms/edit
2023-09-20 13:05:00,669 [jetty-183 rqc:1 owner oQdVKBYp] INFO  rems.middleware - wrap-entry-logging {}
2023-09-20 13:05:00,682 [jetty-183 rqc:1 owner oQdVKBYp] INFO  rems.middleware - > :put /api/forms/edit lang: :en user: {:userid owner, :name Owner, :email owner@example.com}  roles: #{:logged-in :owner}
2023-09-20 13:05:00,682 [jetty-183 rqc:1 owner oQdVKBYp] INFO  rems.middleware - wrap-context-logging {}
2023-09-20 13:05:00,690 [jetty-183 rqc:1 owner oQdVKBYp] INFO  rems.api.util - extended-logging /api/forms/edit {:organization {:organization/id hus}, :form/internal-name dsdsssee, :form/external-title {:en sd, :fi sdsd, :sv sdsd}, :form/fields [], :form/id 14}
2023-09-20 13:05:00,700 [jetty-183 rqc:1 owner oQdVKBYp] INFO  rems.middleware - req < :put /api/forms/edit 200 30ms
```

Where the first `wrap-context-logging`/`wrap-entry-logging` correspond to the following two additional lines
```diff clojure
(defn wrap-context-logging
  "Logs detailed information about a request just before
  the main business code is run.

  Here we have the most context about the user and request available."
  [handler]
  (fn [request]
    (let [uri (str (:uri request)
                   (when-let [q (:query-string request)]
                     (str "?" q)))
          log? (not (re-matches silence-logging-regex uri))]
      (when log?
        (log/info ">" (:request-method request) uri
                  "lang:" context/*lang*
                  "user:" context/*user*
                  (if (:uses-valid-api-key? request)
                    "api-key"
                    "")
                  "roles:" context/*roles*)

        (log/debug "session" (pr-str (:session request)))
+       (log/info "wrap-context-logging" (pr-str (:params request)))
        (when (seq (:form-params request))
          (log/debug "form params" (pr-str (:form-params request)))))

      (handler request))))

(defn wrap-entry-logging
  "Logs every request into an entry `req >` and exit `req <` log.

  This happens in the very beginning of the request processing, so
  not a lot of context is available yet.

  This is important to show a request as soon as it's possible, so if it
  starts to wait on middleware DB requests or such, it still appears
  in the logs."
  [handler]
  (fn [request]
    (let [uri (str (:uri request)
                   (when-let [q (:query-string request)]
                     (str "?" q)))
          start (System/nanoTime)
          log? (not (re-matches silence-logging-regex uri))
          userid (or (get-in request [:session :identity :userid])
                     (:userid context/*user*)
                     (auth/get-api-userid request))]
      (try
        (swap! request-count inc)

        (with-mdc {:userid userid
                   :request-count @request-count}
          (when log?
            (log/info "req >" (:request-method request) uri)
+           (log/info "wrap-entry-logging" (pr-str (:params request)))
            (when (seq (:form-params request))
              (log/debug "form params" (pr-str (:form-params request)))))

          (let [response (handler request)]
            (when log?
              (log/info "req <" (:request-method request) uri (:status response)
                        (str (int (/ (- (System/nanoTime) start) 1000000)) "ms")
                        (or (get-in response [:headers "Location"]) "")))
            response))
        (finally
          (swap! request-count dec))))))
```

Do you have a suggestion for a better place for this?

Reply:
I can have a look at the approach and confirm that it should work. The thing with the middleware wrappers is that they must be placed in the correct order, so at least if we want to use the `:params`, then we must place it after the standard params wrapper has done its job. Maybe there is something else too.

Reply:
Another problem that may happen is that one needs to restart the server for changes in middleware to take effect.

But I guess the main problem is that the request body contains most parameters and in a typical request (when they are not actually as HTTP form data) it's not available through params. So the solution should log both params and the body.

Then, about reading the body twice. In another project I implemented my own params reader middleware (just patching the code of the default one), because that was the only place that consumed the body from the stream, or complained about it being read twice. It's something to think about here. It depends on which middleware here actually wants to read the body, and how easily it can be changed.

Likely it's a little bit more involved to read the body separately to log it, but I'm sure it can be made to work.

Reply:
> I can have a look at the approach and confirm that it should work. The thing with the middleware wrappers is that they must be placed in the correct order, so at least if we want to use the `:params`, then we must place it after the standard params wrapper has done its job. Maybe there is something else too.

Thank you, that would be a great help. This is my first time writing Clojure, so my understanding is still quite limited.

Reply:
Though I am wondering, as described in [this comment](https://github.com/CSCfi/rems/pull/3187#discussion_r1333003179), whether it is desirable to do extended logging for all requests, which I don't think it is.

And in that case, there is going to be some code that is specific to those routes somewhere anyways.

The other option is to write generalized code that catches the mutable entity cases, but I think that is going to add unnecessary complexity, as opposed to being careful when new entities and create/edit routes are added, whether or not to add the extended logging functionality as well.

Reply:
I will change this later so that every endpoint is caught automatically. For example I don't see applications (attachments), blacklist or email here, which I would like to be included for completeness. It's easy to forget to add this feature in the future too.